### PR TITLE
workload/schemachange: concurrent randomized schema change workload

### PR DESCRIPTION
**Pull request by spaskob**
Randomly generate (concurrent) schema changes. The tables intentionally
contain no actual data as the focus here is on stressing the machinery
around schema changes, not the machinery around backfills.

Touches #23286.

Release note: None

Release justification: non-production code changes. This is only test
code.

### CODE REVIEW

**Review by petermattis — COMMENTED** [review]
Thanks for picking this up, @spaskob.

You should remove the `[WIP]` from the commit messages.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-:-M3N7O8P-iqlk0G3X_lB:b-8tzt5d)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @spaskob)

___
*[pkg/workload/schemachange/schemachange.go, line 303 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3N3Om--WBYBSBR-_X5:-M3N3Om--WBYBSBR-_X6:bifdlpk) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ae32081a1be5fc8893fb11cb1e394af444b791e/pkg/workload/schemachange/schemachange.go#L303)):*
> ```Go
> 	}
> 	// TODO(spaskob): figure out what to do with the error returned.
> 	_ := w.runOps(tx, ops)
> ```

Generating the ops to run and then running is a bit different than the previous behavior. With the previous code, the second op could discover a table/column/index that the first op had created. What motivated this part of the change?
___
*[pkg/workload/schemachange/schemachange.go, line 741 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3N6bW0DdhUsrtlhABp:-M3N6bW0DdhUsrtlhABq:b-5gu23l) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ae32081a1be5fc8893fb11cb1e394af444b791e/pkg/workload/schemachange/schemachange.go#L741)):*
> ```Go
> ) (name string, err error) {
> 	r := w.rng.Intn(100)
> 	if r < pctExisting {
> ```

There is a general preference to avoid named return values except when they are necessary. I have a personal preference to early return when possible. I think we could achieve both by doing:

```
if w.rng.Intn(100) >= pctExisting {
  return fmt.Sprintf(...), err
}

// Here goes the previous code.
```

**Review by spaskob — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46632)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @petermattis)

___
*[pkg/workload/schemachange/schemachange.go, line 303 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3N3Om--WBYBSBR-_X5:-M3TItXV-R-s849dQT28:b6wujhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ae32081a1be5fc8893fb11cb1e394af444b791e/pkg/workload/schemachange/schemachange.go#L303)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Generating the ops to run and then running is a bit different than the previous behavior. With the previous code, the second op could discover a table/column/index that the first op had created. What motivated this part of the change?
</blockquote></details>

I missed that! The error handling in `run` is subtle and that was a way to wrap my head around it.
I briefly chatted to @ajwerner yesterday about this workload and he had some requests around better reporting of errors. Here are some different types of errors that can occur:

1. from executing an operation.
2. from `BEGIN`, `COMMIT` and `ROLLBACK`
3. from `randOp` 

Additionally, the logic around what is printed and in what order was also a bit hard to follow.
I've tried a different refactoring approach (see new commit) - let me know how you like it. I hope it also makes it easier to write a test if we decided one is needed.

We need to decide from workload's users point of view how to communicate the different types of errors and when error should abort the workload. @ajwerner could you please chime in as well.
___
*[pkg/workload/schemachange/schemachange.go, line 741 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3N6bW0DdhUsrtlhABp:-M3TJJrbAzqNHJ8zHHk-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ae32081a1be5fc8893fb11cb1e394af444b791e/pkg/workload/schemachange/schemachange.go#L741)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

There is a general preference to avoid named return values except when they are necessary. I have a personal preference to early return when possible. I think we could achieve both by doing:

```
if w.rng.Intn(100) >= pctExisting {
  return fmt.Sprintf(...), err
}

// Here goes the previous code.
```
</blockquote></details>

Done.

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46632)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @spaskob)

___
*[pkg/workload/schemachange/schemachange.go, line 303 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3N3Om--WBYBSBR-_X5:-M3WaOlC53A1w2e7zM8m:b-jvibh2) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ae32081a1be5fc8893fb11cb1e394af444b791e/pkg/workload/schemachange/schemachange.go#L303)):*
> The error handling in run is subtle and that was a way to wrap my head around it.

Heh. It sort of grew organically as the PR evolved. I agree it had reach a somewhat con
___
*[pkg/workload/schemachange/schemachange.go, line 741 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3N6bW0DdhUsrtlhABp:-M3W_Ivs1njeLi8umpqG:b-we3cml) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ae32081a1be5fc8893fb11cb1e394af444b791e/pkg/workload/schemachange/schemachange.go#L741)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Looks good. Let's follow this pattern for the other `rand*` methods.
___
*[pkg/workload/schemachange/schemachange.go, line 207 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3WamNA3Y53bMN3rzZ1:-M3WamNA3Y53bMN3rzZ2:bg71py6) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca49468782b5f4efc407f717ac1ad69aac58bb43/pkg/workload/schemachange/schemachange.go#L207)):*
> ```Go
> // used to generate new unique names. Note that this assumes that no
> // other workload is being run at the same time.
> // TODO(spaskob): Enable workloads running concurrently.
> ```

Can you elaborate on what you're think of here? I suspect one `workload` instance can drive enough schema change concurrency for even a very large cluster.
___
*[pkg/workload/schemachange/schemachange.go, line 239 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3WahXs5jtvoAlrviW2:-M3WahXs5jtvoAlrviW3:b-34k0l2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca49468782b5f4efc407f717ac1ad69aac58bb43/pkg/workload/schemachange/schemachange.go#L239)):*
> ```Go
> }
> 
> func (w *schemaChangeWorker) runOps(tx *pgx.Tx, ops []string) (err error) {
> ```

I believe this method is now unused and can be deleted.
___
*[pkg/workload/schemachange/schemachange.go, line 306 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3Wa1Eq5B7upD6BNe9x:-M3Wa1Eq5B7upD6BNe9y:b-h5g6rm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca49468782b5f4efc407f717ac1ad69aac58bb43/pkg/workload/schemachange/schemachange.go#L306)):*
> ```Go
> }
> 
> func (w *schemaChangeWorker) wrapInTxn(_ context.Context) error {
> ```

I'm ok with this structure, though this name doesn't look quite right to me. We're not actually wrapping any of the arguments. How about naming this method `run`, and renaming the existing `run` method to `runInTxn`?

**Review by spaskob — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46632)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @petermattis)

___
*[pkg/workload/schemachange/schemachange.go, line 207 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3WamNA3Y53bMN3rzZ1:-M3X-SfH08h4PFf9Xech:b6r1uvl) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca49468782b5f4efc407f717ac1ad69aac58bb43/pkg/workload/schemachange/schemachange.go#L207)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Can you elaborate on what you're think of here? I suspect one `workload` instance can drive enough schema change concurrency for even a very large cluster.
</blockquote></details>

sorry bad choice of words, I am worried about the workload semantics if for whatever reason two workloads are run on the same cluster. It's possible that I am seeing ghosts.
___
*[pkg/workload/schemachange/schemachange.go, line 239 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3WahXs5jtvoAlrviW2:-M3X-dlm8uh4xemcNR4I:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca49468782b5f4efc407f717ac1ad69aac58bb43/pkg/workload/schemachange/schemachange.go#L239)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I believe this method is now unused and can be deleted.
</blockquote></details>

Done.
___
*[pkg/workload/schemachange/schemachange.go, line 306 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3Wa1Eq5B7upD6BNe9x:-M3X-h4XA_0XZ0VTKaeR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca49468782b5f4efc407f717ac1ad69aac58bb43/pkg/workload/schemachange/schemachange.go#L306)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm ok with this structure, though this name doesn't look quite right to me. We're not actually wrapping any of the arguments. How about naming this method `run`, and renaming the existing `run` method to `runInTxn`?
</blockquote></details>

Done.

**Review by petermattis — APPROVED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-:-M3aYqFjCZeHis8hfU6x:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @spaskob)

___
*[pkg/workload/schemachange/schemachange.go, line 207 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3WamNA3Y53bMN3rzZ1:-M3aY64z2-u_8We2KMhO:b-svawys) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca49468782b5f4efc407f717ac1ad69aac58bb43/pkg/workload/schemachange/schemachange.go#L207)):*
<details><summary><i>Previously, spaskob (Spas Bojanov) wrote…</i></summary><blockquote>

sorry bad choice of words, I am worried about the workload semantics if for whatever reason two workloads are run on the same cluster. It's possible that I am seeing ghosts.
</blockquote></details>

Got it. Running multiple instances of the same workload on a cluster is pretty rare. We do it for geo-partitioned tpcc and perhaps one or two other tests. Most of the time we run a single workload instance. I think that will be true of this workload. That said, I'm fine with leaving this TODO here.
___
*[pkg/workload/schemachange/schemachange.go, line 72 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3aXl8w3NY29dJn82_a:-M3aXl8w3NY29dJn82_b:b-xv2hfx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d2b0773519d7b89cb535f27ac1bceaa5923d323/pkg/workload/schemachange/schemachange.go#L72)):*
> ```Go
> 		s.flags.StringVar(&s.dbOverride, `db`, ``,
> 			`Override for the SQL database to use. If empty, defaults to the generator name`)
> 		s.flags.IntVar(&s.concurrency, `concurrency`, 2*runtime.NumCPU(), /* TODO(spaskob): sensible default? */
> ```

I think the sensible default will change over time. `2` seems too low. `2*num-cpu` seems high. Fine with leaving this as-is for now.
___
*[pkg/workload/schemachange/schemachange.go, line 171 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3aY-5rB08XhXewRB0o:-M3aY-5rB08XhXewRB0p:bt6alwr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d2b0773519d7b89cb535f27ac1bceaa5923d323/pkg/workload/schemachange/schemachange.go#L171)):*
> ```Go
> 	}
> 	cfg := workload.MultiConnPoolCfg{
> 		MaxTotalConnections: s.concurrency * 2, //TODO
> ```

Nit: `s/TODO/TODO(spaskob)`
___
*[pkg/workload/schemachange/schemachange.go, line 289 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3aY_zj899ho6A443AP:-M3aY_zj899ho6A443AQ:b7i7kh6) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d2b0773519d7b89cb535f27ac1bceaa5923d323/pkg/workload/schemachange/schemachange.go#L289)):*
> ```Go
> 		}
> 	}
> 	// TODO(spaskob): what to do with the error instead dropping it?
> ```

Some workloads have a `--tolerate-errors` flag. I didn't add one to this workload because errors are an essential part of the workload. I think logging them is fine.

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46632)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @spaskob)

___
*[pkg/workload/schemachange/schemachange.go, line 303 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3N3Om--WBYBSBR-_X5:-M3gDwbS0dprDYp5EtbW:by5hpdj) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ae32081a1be5fc8893fb11cb1e394af444b791e/pkg/workload/schemachange/schemachange.go#L303)):*
> We need to decide from workload's users point of view how to communicate the different types of errors and when error should abort the workload. @ajwerner could you please chime in as well.

Yes indeed. The long tail work on this workload is to introspect errors and determine whether they're expected or unexpected. Can we add a big TODO about that before merging this? Eventually tolerate errors should be added to tollerate unexpected errors and then unexpected errors should fail the workload.

For example, an attempt to do something we don't support should be swallowed (though if we can detect that maybe we should just not do it, e.g). It will be hard to use this test for anything more than liveness detection until we go through the tedious process of classifying errors.

**Review by spaskob — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46632)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @petermattis)

___
*[pkg/workload/schemachange/schemachange.go, line 303 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3N3Om--WBYBSBR-_X5:-M3gQdph5acn9agFlYqT:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ae32081a1be5fc8893fb11cb1e394af444b791e/pkg/workload/schemachange/schemachange.go#L303)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

> We need to decide from workload's users point of view how to communicate the different types of errors and when error should abort the workload. @ajwerner could you please chime in as well.

Yes indeed. The long tail work on this workload is to introspect errors and determine whether they're expected or unexpected. Can we add a big TODO about that before merging this? Eventually tolerate errors should be added to tollerate unexpected errors and then unexpected errors should fail the workload.

For example, an attempt to do something we don't support should be swallowed (though if we can detect that maybe we should just not do it, e.g). It will be hard to use this test for anything more than liveness detection until we go through the tedious process of classifying errors. 
</blockquote></details>

done
___
*[pkg/workload/schemachange/schemachange.go, line 207 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3WamNA3Y53bMN3rzZ1:-M3gQXmRBhm4hJ_sWLue:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca49468782b5f4efc407f717ac1ad69aac58bb43/pkg/workload/schemachange/schemachange.go#L207)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Got it. Running multiple instances of the same workload on a cluster is pretty rare. We do it for geo-partitioned tpcc and perhaps one or two other tests. Most of the time we run a single workload instance. I think that will be true of this workload. That said, I'm fine with leaving this TODO here.
</blockquote></details>

Done.
___
*[pkg/workload/schemachange/schemachange.go, line 72 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3aXl8w3NY29dJn82_a:-M3gQWD18Cusv9CtoZk8:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d2b0773519d7b89cb535f27ac1bceaa5923d323/pkg/workload/schemachange/schemachange.go#L72)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I think the sensible default will change over time. `2` seems too low. `2*num-cpu` seems high. Fine with leaving this as-is for now.
</blockquote></details>

Done.
___
*[pkg/workload/schemachange/schemachange.go, line 289 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/46632#-M3aY_zj899ho6A443AP:-M3gQ_xI1Vw5KMfgZXpL:b-3rrbu3) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d2b0773519d7b89cb535f27ac1bceaa5923d323/pkg/workload/schemachange/schemachange.go#L289)):*
>  fine.
added todo

**Review by petermattis — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46632)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)
